src/Lucene.Net/Search/CachingWrapperFilter.cs

```diff
@@ -79,68 +81,65 @@ public FilterCache(DeletesMode deletesMode)
-                    }
-                    else if (deletesMode == DeletesMode.RECACHE)
-                    {
-                        // key on deletes, if any, else core
-                        value = cache[delCoreKey];
-                    }
-                    else
-                    {
+                if (deletesMode == DeletesMode.IGNORE)
+                {
+                    // key on core
+                    current.TryGetValue(coreKey, out value);
+                }
+                else if (deletesMode == DeletesMode.RECACHE)
+                {
+                    // key on deletes, if any, else core
+                    current.TryGetValue(delCoreKey, out value);
+                }
+                else
+                {
 
-                        System.Diagnostics.Debug.Assert(deletesMode == DeletesMode.DYNAMIC);
+                    System.Diagnostics.Debug.Assert(deletesMode == DeletesMode.DYNAMIC);
 
-                        // first try for exact match
-                        value = cache[delCoreKey];
+                    // first try for exact match
+                    current.TryGetValue(delCoreKey, out value);
 
-                        if (value == null)
+                    if (value == null)
+                    {
+                        // now for core match, but dynamically AND NOT
+                        // deletions
+                        current.TryGetValue(coreKey, out value);
+                        if (value != null && reader.HasDeletions)
                         {
-                            // now for core match, but dynamically AND NOT
-                            // deletions
-                            value = cache[coreKey];
-                            if (value != null && reader.HasDeletions)
-                            {
-                                value = MergeDeletes(reader, value);
-                            }
+                            value = MergeDeletes(reader, value);
                         }
                     }
-                    return value;
                 }
 
+                return value;
+
             }
        
             protected abstract T MergeDeletes(IndexReader reader, T value);
 
             public void Put(object coreKey, object delCoreKey, T value)
             {
-                lock (this)
+                if (deletesMode == DeletesMode.IGNORE)
                 {
-                    if (deletesMode == DeletesMode.IGNORE)
-                    {
-                        cache[coreKey] = value;
-                    }
-                    else if (deletesMode == DeletesMode.RECACHE)
-                    {
-                        cache[delCoreKey] = value;
-                    }
-                    else
-                    {
-                        cache[coreKey] = value;
-                        cache[delCoreKey] = value;
-                    }
+                    _cache.Add(coreKey, value);
```

Comment:
What happens if the value already exists here? Does it overwrite?

Reply:
`Add` will throw.. will use `AddOrUpdate`